recipes/lora_finetune_single_device.py

```diff
@@ -491,20 +515,54 @@ def save_checkpoint(self, epoch: int) -> None:
             )
 
         # Move to CPU to avoid a copy on GPU
-        state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
+        # state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
+
+        # === Create a FakeTensor version of the state_dict (preserves storage aliasing) ===
+        state_dict = self._model.state_dict()
+        mode = FakeTensorMode()
+        converter = FakeTensorConverter()
+        fake_state_dict = OrderedDict()
+        for k, v in state_dict.items():
+            # remove adapters (we don't want them in merged_state_dict)
+            if "lora_a" in k or "lora_b" in k:
+                continue
+            if isinstance(v, NF4Tensor):
+                fake_state_dict[k] = converter.from_real_tensor(mode, v).to(torch.bfloat16).cpu()
+            else:
+                fake_state_dict[k] = converter.from_real_tensor(mode, v).cpu()
+
+        # === Save and load fake state_dict ===
+        # dest_state_dict is now backed by an mmap-ed file that can be written to
+        dest_state_dict_path = "/tmp/llama_state_dict.pt"
+        torch.save(fake_state_dict, dest_state_dict_path)
+        torch.serialization.set_default_mmap_options(MAP_SHARED)
+        dest_state_dict = torch.load(dest_state_dict_path, mmap=True, weights_only=True)
+        torch.serialization.set_default_mmap_options(MAP_PRIVATE)
```

Comment:
Maybe worth doing this as a context manager or something. Also noob q: why are we able to put back into `MAP_PRIVATE` mode before we actually copy `dest_state_dict`?

Reply:
Good point, I should add a context manager for this, this is a global config that sets `MAP_SHARED` for all torch.load calls (the implementation is like this https://github.com/pytorch/pytorch/blob/main/torch/serialization.py#L1245-L1251), so changing back to `MAP_PRIVATE`  would not affect the preceding torch.load call